README.md
kiosk-browser
=============

Ubuntu/Debian/Raspbian package to set up a system as a kiosk browser. Release announcement and example photo can be found on my blog at http://blog.schlomo.schapiro.org/2012/09/dashboards-made-easy.html.

This package disables the regular GUI and installs a browser-only GUI. Keyboards and mice are disabled to prevent tampering.

The behavior can be customized in `/etc/default/kiosk-browser`:
*   Set `KIOSK_BROWSER_START_PAGE` to set the start page(s).
*   Set `KIOSK_BROWSER_PORTS` to autoconfigure these|this display ports (Use xrandr port names)
*   Set `KIOSK_BROWSER_XRANDR_EXTRA_OPTS` to rotate some displays or set other custom xrandr settings.
*   Set `KIOSK_BROWSER_WATCHDOG_TIMEOUT` to the amount of seconds after which the systems reboots if the screen did not change. Default is 3600. 
*   Set `KIOSK_BROWSER_WATCHDOG_CHECK_INTERVAL` to the check interval in seconds, default is 313.
*   Set `KIOSK_BROWSER_SHOW_SYSTEM_MONITOR=yes` to show [xosview](http://xosview.sourceforge.net/).
*   Set `KIOSK_BROWSER_VNC_VIEWER_DISPLAY=0` to enable a vncviewer in listening mode on port 5500.
*   Add custom initialization commands or pull the above configuration from somewhere else.

The package is licensed under the GNU Public License, see included
LICENSE.txt for full license text.

Building
========

1. Install `fakeroot`, `lintian` and `dpkg`.
1. Checkout this github repo and type `make`.
1. In `out` you should find the resulting .deb package file.

Installation
============

1. Install the .deb package you build just now from the `out` directory. Alternatively you can download it from the Releases of this GitHub repo or add my [PPA](https://launchpad.net/~sschapiro/+archive/ubuntu/ppa) to your system and install it from there.
1. Configure SSH access so that you don't lock yourself out of your computer
1. Create `/etc/default/kiosk-browser` and set `KIOSK_BROWSER_START_PAGE`
1. Reboot the system

If you want to play with the settings without rebooting the system you can use this command to restart the kiosk browser:
    
    sudo kiosk-browser-control restart


Notes
=====

*   This package disables all inputs in the X server so that nobody can mess with your system or use it as an entry point into your network.


Customisation & Special Features
================================

Multi-Monitor Support
---------------------

kiosk-browser supports setting up multiple monitors with different browser windows. The implementation is somewhat tricky so that I would be happy to get some feedback on it.

Some of the above mentioned `KIOSK_BROWSER_*` variables can be [Bash Arrays](http://tldp.org/LDP/abs/html/arrays.html) and multi-monitor support is enabled by setting these variables to arrays. In each array the same position refers to the same display, for example:

    KIOSK_BROWSER_PORTS=(HDMI1 HDMI2)
    KIOSK_BROWSER_XRANDR_EXTRA_OPTS=("" "--rotate left")
    KIOSK_BROWSER_START_PAGE=( 
        http://blog.schlomo.schapiro.org
        http://go.schapiro.org/schlomo
    )

Note the empty (`""`) array value which means no extra xrandr options for HDMI1. For single monitor operations you can still use these variables to configure the display, e.g.:

    KIOSK_BROWSER_PORTS=VGA
    KIOSK_BROWSER_XRANDR_EXTRA_OPTS="--rotate left"
    KIOSK_BROWSER_START_PAGE=http://blog.schlomo.schapiro.org

If you don't know if the VGA port is called VGA or VGA1 you can specify a substring and the scripts will use the first connected port that matches. Here we basically use the fact that a Bash String is exactly the same as a Bash Array with a single value in it.


Different Browsers
------------------

The default browser is [Chromium](http://www.chromium.org/). kiosk-browser also supports [Epiphany](https://wiki.gnome.org/Apps/Web) as browsers. To use a different browser set the `KIOSK_BROWSER_PROGRAM` variable to either `epiphany`. Since September 2014 epiphany [has been greatly improved on Raspberry Pi](http://www.raspberrypi.org/web-browser-released/) so that it is worth to try it out.

Virtual Projector
-----------------

If you use the kiosk-browser package to drive a large-screen team dashboard then you can also use it as a projector. Enable VNC support with the `KIOSK_BROWSER_VNC_VIEWER_DISPLAY` and "beam" your desktop to the screen with a VNC reverse connection (VNC server connects to the VNC client).

**Linux Clients**

Install `x11vnc` and use `x11vnc -connect <HOSTNAME>` to send your desktop. [x11vnc](http://www.karlrunge.com/x11vnc/) has many options to fine-tune the VNC session, including scaling your desktop to fit the native resolution of the kiosk-browser screen.

**Windows Clients**

Install [TightVNC Server](http://www.tightvnc.com/download.php) and use the "Attach Listening Viewer" menu to connect.

**Mac Clients**

Install [Vine Server (osxvnc)](http://downloads.testplant.com/downloads/Vine/VineServer4.01.dmg) and add a "Reverse Connection"


Chromium Preferences
--------------------

The chromium profile is wiped on each start. To set chromium preferences you can use the master_preferences mechanism and create
a `/usr/lib/chromium-browser/master_preferences` file with preferences. For example I use

    {
        "profile" : {
            "default_zoom_level" : 2.22390108574155
        }
    }

to set the zoom level to 150%. This allows us to have dashboards that look fine both on my desktop and on the kiosk browser.

Show more than one page
-----------------------

Sometimes a dashboard should cycle between several views. I think that this should be done on the server side, not the client side. For example, create on your web server a file named `urls.js` like this with the URLs in it:

        var urls = [ 
                "http://www.schapiro.org/schlomo/publications",
                "http://www.schapiro.org/schlomo/videos"
                ];

And next to it another file with the HTML code, named `dashboard.html`:

        <!DOCTYPE html>
        <html>
        <head>
        <script type="text/javascript" src="urls.js"></script>
        <script type="text/javascript">
        
        function start() {
                setInterval(function(){cycle()},60000);
                cycle();
        }
        
        var counter = 0;
        function cycle() {
                var iframe = document.getElementById("iframe");
                iframe.src=urls[counter++];
                if (counter >= urls.length) {
                        counter = 0;
                }
        }
        
        </script>
        <style type="text/css">
        body,iframe {
                padding: 0px;
                margin: 0px;
        }
        #wrap { position:fixed; left:0; width:100%; top:0; height:100%; }
        #iframe { display: block; width:100%; height:100%; }
        </style>
        <title>Dashboard switcher by Schlomo Schapiro</title>
        </head>
        <body onload="start()">
        <div id="wrap"><iframe id="iframe" src=""/></div>
        </body>
        </html>

Finally, adjust `KIOSK_BROWSER_START_PAGE` in `/etc/default/kiosk-browser` to point to this `dashboard.html` and you are done.

Prevent loading
---------------

As `kiosk-browser` disables mouse and keyboard it can be nearly impossible to fix network connectivity problems. If you want to be able to prevent the loading of the graphical environment create `/etc/systemd/system/nodm.service.d/keycheck.conf` with the this content:
```
[Service]
ExecStartPre=/bin/sh -c "timeout 1 thd --dump /dev/input/event* | grep -q _ && exit 1 || exit 0"

```
As a result pressing any key or mouse button during system startup will prevent the start of `nodm` and the graphical environment. This trick requires the `triggerhappy` package which is installed by default on Raspbian.

Hacking
=======

The kiosk-browser user session creates a temporary Home Directory under `/tmp`. To access the running kiosk-browser session use `kiosk-browser-control interactive`. In the shell you can use `cd` to go to the session Home Directory.


debian/changelog
kiosk-browser (107) stable; urgency=medium

  * Tell people to setup remote access before installing
  * fix releases link
  * Support systemd to collect network info in a nicer way Remove unwanted lines from traditional network info
  * Use Chrome policy to disable translate old way doesn't work any more

 -- Schlomo Schapiro <schlomo@schapiro.org>  Sat, 16 Feb 2019 20:58:03 +0100

kiosk-browser (102) stable; urgency=medium

  * handle buildinfo
  * Remove uzbl in readme

 -- Schlomo Schapiro <schlomo@schapiro.org>  Tue, 12 Feb 2019 21:09:21 +0100

kiosk-browser (99) stable; urgency=medium

  * Remove uzbl to support Ubuntu 18.10

 -- Schlomo Schapiro <schlomo@schapiro.org>  Tue, 12 Feb 2019 21:01:11 +0100

kiosk-browser (21) stable; urgency=medium

  * Prevent debconf questions

 -- Schlomo Schapiro <schlomo@schapiro.org>  Tue, 23 May 2017 20:37:55 +0200

kiosk-browser (20) stable; urgency=medium

  * Migrate to work with systemctl (based on work from https://github.com/MartinLinSup/kiosk-browser)
  * Include Vagrantfile to test on Ubuntu and on Debian
  * Resume development, see GitHub for older changes

 -- Schlomo Schapiro <schlomo.schapiro@zalando.de>  Mon, 22 May 2017 16:53:18 +0200


xsession.sh
#!/bin/bash
exec &> >(exec logger -t "kiosk-browser[$$]" )

# sanitize environment
export LANG=en_US LANGUAGE=en_US LC_ALL=en_US.UTF-8
unset ${!XDG_*}
set -x

KILL_ON_EXIT=
HOME=$( mktemp -d /tmp/kiosk-browser.HOME.XXXXXXXXXXXXX )
function exittrap {
    procs=( $KILL_ON_EXIT $(jobs -p))
    kill -TERM ${procs[@]}
    (
        # wait in background for all procs to be gone before removing the old home directory
        # reason is that nodm KILLs the session 2 seconds after TERMinating it
        sleep 5 &
        wait $!
        kill -KILL $KILL_ON_EXIT ${procs[@]} &>/dev/null
        rm -Rf $HOME
        pstree -paul $USER
        pkill -KILL -e -u $USER # harakiri
    ) </dev/null &
}
trap exittrap TERM EXIT # kill subprocesses on exit

# dump what we got so far
declare -p
pstree -paul $USER

# Wait 30 seconds for network by checking for default route
(( end_time=SECONDS+30 ))
until (( SECONDS > end_time)) || [[ "$(ip route)" == *default* ]] ; do
    sleep 2 &
    wait $!
done

# show debug info for 60 seconds as overlay
{
    echo "Welcome to the Kiosk Browser (http://github.com/schlomo/kiosk-browser)"
    echo
    echo "This is $(uname -n)"
    echo
    if type -p networkctl &>/dev/null ; then
        networkctl status
    else
        ip a | grep -E 'UP|inet'
        echo
        ip route
        echo
        sed -e '/^#/d;/^$/d' /etc/resolv.conf
    fi
    echo
    perl -e '$/ = undef; $d=<>; $d =~ m/.*(lease \{.*?\})$/s ; print $1' $(ps ax | grep dhclient | sed -ne "s/.* \(\/[^ ]\+\.lease[s]\?\).*/\1/p") <<<""
    echo
    echo "This message will self-destruct in 60 seconds"
} | sed -e 's/●/ /g' | osd_cat --offset 10 --indent 10 --colour green --outline 2 --font 10x20 --lines 50 --delay 60 &

disown -a # forget about running osd_cat, it will terminate itself anyway

# window manager helps with fullscreen, window manager must support XINERAMA for multi-screen setups
openbox --debug --config-file /usr/share/kiosk-browser/openbox-rc.xml &


# cache xrandr configuration
XRANDR_OUTPUT="$(xrandr)"
function xrandr_find_port {
    # find connected port matching pattern
    sed -n -e "/$1.* connect/s/^\([^ ]\+\).*/\1/p" <<<"$XRANDR_OUTPUT"
}

if test -r /etc/default/kiosk-browser ; then
    source /etc/default/kiosk-browser
fi


if [[ ! "$KIOSK_BROWSER_PORTS" ]] ; then
    # set . as a built-in default to use the first connected port that xrandr reports
    KIOSK_BROWSER_PORTS=.
fi

if [[ ! "$KIOSK_BROWSER_START_PAGE" ]] ; then
    # point to our github page as built-in default
    KIOSK_BROWSER_START_PAGE=https://github.com/schlomo/kiosk-browser
fi

KIOSK_BROWSER_WATCHDOG_CHECK_INTERVAL="${KIOSK_BROWSER_WATCHDOG_CHECK_INTERVAL:-313}"
if [[ ! "$KIOSK_BROWSER_WATCHDOG_TIMEOUT" ]] ; then
    # default stale screen watchdog is 1h
    KIOSK_BROWSER_WATCHDOG_TIMEOUT=3600
fi

if [[ ! "$KIOSK_BROWSER_VNC_VIEWER_DISPLAY" ]] ; then
    # disable VNC viewer by default
    KIOSK_BROWSER_VNC_VIEWER_DISPLAY=-1
fi

# configure displays
xrandr $(
    xrandr_position=
    for (( c=0 ; c<${#KIOSK_BROWSER_PORTS[@]} ; c++ )) ; do
        port=$(xrandr_find_port "${KIOSK_BROWSER_PORTS[c]}")
        if [[ "$port" ]] ; then
            #                                                                        / we use xrandr_position as indicator of first port
            echo "--output $port ${KIOSK_BROWSER_XRANDR_EXTRA_OPTS[c]} $( [[ -z "$xrandr_position" ]] && echo --primary ) $xrandr_position --auto"
            xrandr_position="--right-of $port"
        else
            : skipping configuration of "${KIOSK_BROWSER_PORTS[c]}" because it is not connected
        fi
    done
)

# wait a moment for displays to settle
sleep 5 &
wait $!

# xrandr configuration changed, update cache
XRANDR_OUTPUT="$(xrandr)"

# disable screen blanking
xset -dpms
xset s off
xset s noblank

# start watchdog, reboot system if screen stops to change
if (( KIOSK_BROWSER_WATCHDOG_TIMEOUT > 0 )) ; then
    (
	WATCHDOG_PID=$BASHPID
	exec &> >(exec logger -t "kiosk-browser[$$]-watchdog[$WATCHDOG_PID]" )
	: "Starting watchdog. Timeout is $KIOSK_BROWSER_WATCHDOG_TIMEOUT seconds, checking every $KIOSK_BROWSER_WATCHDOG_CHECK_INTERVAL seconds"
        trap 'kill -TERM $(jobs -p)' TERM

        LASTHASH=""
        LASTCHANGED="$SECONDS"
        while sleep $KIOSK_BROWSER_WATCHDOG_CHECK_INTERVAL & wait $! ; do
            HASH=$(nice import -display :0 -window root -monochrome jpg:- | nice identify -format '%#' -)
            #declare -p HASH LASTHASH LASTCHANGED SECONDS
            if [[ "$HASH" = "$LASTHASH" ]] ; then
                if (( SECONDS > LASTCHANGED + KIOSK_BROWSER_WATCHDOG_TIMEOUT )) ; then
                sudo /sbin/reboot
                break
                fi
            else
                LASTHASH="$HASH"
                LASTCHANGED="$SECONDS"
            fi
        done
    ) </dev/null 1>&2 &
fi

if [[ "$KIOSK_BROWSER_SHOW_SYSTEM_MONITOR" ]] ; then
    xosview &
fi

# start vnc viewer if requested
if [[ "$KIOSK_BROWSER_VNC_VIEWER_DISPLAY" ]] && (( KIOSK_BROWSER_VNC_VIEWER_DISPLAY >= 0 )) ; then
    vncviewer -fullscreen -viewonly -listen "$KIOSK_BROWSER_VNC_VIEWER_DISPLAY" &
fi

# remember system jobs
KILL_ON_EXIT=$(jobs -p)
# the wait below should wait only for the browsers and not hang on the system jobs
disown -a

# Ubuntu has chromium-browser and Debian wheezy has chromium
CHROME=$(type -p chromium-browser 2>/dev/null)
if [[ -z "$CHROME" ]] ; then
    CHROME=$(type -p chromium 2>/dev/null)
fi


UZBL_CONFIG="set show_status=0
set geometry=maximized"

while sleep 5 & wait $!; do
    # exit if no display given, use xwininfo to test for running X server
    xwininfo -root &>/dev/null || exit 99

    # if KIOSK_BROWSER_PORTS is set, assume that it specifies multiple screens connected.
    for (( c=0 ; c<${#KIOSK_BROWSER_PORTS[@]} ; c++ )) ; do
        port=$(xrandr_find_port "${KIOSK_BROWSER_PORTS[c]}")
        [[ -z "$port" ]] && continue # skip for all invalid ports (that are not a number)
        port_x=$(sed -ne "/$port/s#[^+].*+\([0-9]\+\)+.*#\1#p" <<<"$XRANDR_OUTPUT")

        BROWSER_PROFILE_DIR=~/profile-$c
        URL="${KIOSK_BROWSER_START_PAGE[c]:-$KIOSK_BROWSER_START_PAGE}"

        # start with empty profile dir
        [[ -d $BROWSER_PROFILE_DIR ]] && rm -Rf $BROWSER_PROFILE_DIR
        mkdir -p $BROWSER_PROFILE_DIR

        if [[ "$KIOSK_BROWSER_PROGRAM" == "epiphany" ]] && type -p epiphany-browser &>/dev/null ; then
            epiphany-browser --class epiphany-$c --profile $BROWSER_PROFILE_DIR "$URL" &
            PID=$!
            sleep 10 &
            wait $!
            # Use PID to distinguish browsers
            xdotool search --class epiphany-$c windowmove --sync $port_x 0 key F11
        elif [[ "$KIOSK_BROWSER_PROGRAM" == "uzbl" ]] && type -p uzbl &>/dev/null ; then
            uzbl -n uzbl-$c -c - "$URL" <<<"$UZBL_CONFIG" &
            sleep 5 &
            wait $!
            xdotool search --class uzbl-$c windowmove --sync $port_x 0
        else
            $CHROME --user-data-dir=$BROWSER_PROFILE_DIR "${KIOSK_BROWSER_OPTIONS[@]}" --use-fake-ui-for-media-stream --no-first-run --start-fullscreen --app="$URL" &
            PID=$!

            # move new window to the current screen. We identify the window by the --user-data-dir option which appears in the window class name :-)

            starttime=$SECONDS
            while ! xdotool search --onlyvisible --pid $PID --name any windowmove --sync $port_x 0 ; do
                if (( SECONDS-starttime > 30 )) ; then
                    killall -TERM ${CHROME##*/} # if we can't find a chrome window then we kill them all and try again, killall needs only basename
                    sleep 2 &
                    wait $!
                    killall -KILL ${CHROME##*/}
                    kill -KILL $(jobs -p) # kill all background jobs (= browsers) to restart outer while loop
                    break 2 # abort the for loop
                fi
                sleep 2 &
                wait $!
            done
        fi
    done

    xdotool search xosview windowactivate # make sure xosview is visible

    wait # for the browsers to finish
done

# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
